### PR TITLE
Created basic outline for doubly linked list, with very basic tests

### PR DESCRIPTION
**Pull request by TreverArnold**

### CODE REVIEW

**Comment by TreverArnold** [comment]
(WIP)

File: doubly_linked_list.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+class Node:
+    def __init__(self, value):
+        self.value = value
+        self.next = None
+        self.prev = None
+
+class doubly_LinkedList:
```
Comment (W-Ely) [inline doubly_linked_list.py]: UpperCamelCase

File: doubly_linked_list.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            self.tail = new_node
+            self.tail.prev = None
+            
+        self.head = new_node
+        self.head.prev = None
+        self._size += 1
+
+    def append(self, value):
+        new_node = Node(value)
+        new_node.next = self.tail
```
Comment (W-Ely) [inline doubly_linked_list.py]: Appending adds to the tail. Why would the new Node's next be the current tail?  Shouldn't it be the previous?

File: doubly_linked_list.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    def push(self, value):
+        new_node = Node(value)
+        new_node.next = self.head
+
+        if self.head is not None:
+            self.head.prev = new_node
+        else:
+            self.tail = new_node
+            self.tail.prev = None
```
Comment (W-Ely) [inline doubly_linked_list.py]: What is the current value of new_node.previous? Why is line 25 needed here?

File: doubly_linked_list.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            
+        self.head = new_node
+        self.head.prev = None
```
Comment (W-Ely) [inline doubly_linked_list.py]: What is the current value of self.head.previous? Why is line 28 needed here?

File: doubly_linked_list.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        self._size += 1
+
+    def append(self, value):
+        new_node = Node(value)
+        new_node.next = self.tail
+        self.tail = new_node
+        self._size += 1
+
+        if self._size == 2:
+            self.tail.next = self.head
```
Comment (W-Ely) [inline doubly_linked_list.py]: Unclear to me why the .next of the tail would ever be anything but None?

File: doubly_linked_list.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    def pop(self):
+        if self.head is None:
+            raise ValueError("Pop does not work on an empty list")
+
+        value = self.head.value
+        self.head = self.head.next
+        self._size -= 1
+        return value
```
Comment (W-Ely) [inline doubly_linked_list.py]: I don't see any adjustment of the .tail here.  If you pop the last item from the list the tail should be `None`.

File: doubly_linked_list.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        if self.tail is None:
+            raise ValueError("Shift does not work on an empty list")
+        value = self.tail.value
+        self.tail = self.tail.prev
+        self._size -= 1
+        if self._size == 1:
+            self.tail.next = None
+            self.tail.prev = None
+            self.head.next = None
+            self.head.prev = None
```
Comment (W-Ely) [inline doubly_linked_list.py]: The head's .prev should always be None and the tail's .next, so if all 4 are needed something is off here.

File: doubly_linked_list.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        return value
+
+    def remove(self, node):
+        if self.head is None:
+            raise ValueError("Remove does not affect empty lists")
+
+        if node == self.head:
+            self.head = self.head.next
+            self._size -= 1
+            return
```
Comment (W-Ely) [inline doubly_linked_list.py]: On a list 1 long, this would leave the tail with a node present.

File: doubly_linked_list_test.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    assert my_list._size == 0
+    assert my_list.head == None
+    assert my_list.tail == None
+
+def test_doubly_linked_list_append():
+    my_list = doubly_LinkedList([1, 2])
+    my_list.append(3)
+    assert my_list.tail.value == 3
+    assert my_list.tail.next.value == 1
+    assert my_list._size == 3
```
Comment (W-Ely) [inline doubly_linked_list_test.py]: You should check the the structure comprehensively on every step.  This means my_list.head.... should have it's values and pointer checked to.

File: doubly_linked_list_test.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+def test_doubly_linked_list_shift():
+    my_list = doubly_LinkedList([1, 2])
+    assert my_list.shift() == 1
+    assert my_list.tail.value == 2
+    assert my_list.tail.next == None
+    assert my_list._size == 1
```
Comment (W-Ely) [inline doubly_linked_list_test.py]: Anytime you checkout a .next, a .previous or .previous.value should also be checked.

File: doubly_linked_list_test.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        my_list = doubly_LinkedList()
+        my_list.shift()
+
+def test_doubly_linked_list_push_empty():
+    my_list = doubly_LinkedList([1, 2])
+    my_list.push('')
+    assert my_list._size == 3
+    assert my_list.head.value == ''
+
+def test_doubly_linked_list_push():
```
Comment (W-Ely) [inline doubly_linked_list_test.py:104]: I want to see all methods that manipulate the list, `push`, `pop`, `append`, `shift` to start at 0 or 3 long and go all the way the other direction with the structure comprehensively checked on each step.

Example:
```
dll = DoublyLinkedList([3,2,1])
assert dll.head.prev = None
assert dll.head.value = 1
assert dll.head.next.value = 2
assert dll.head.next.next.value = 3
assert dll.head.next.next.next = None
assert dll.tail.next = None
assert dll.tail.value = 3
assert dll.tail.prev.value = 2
assert dll.tail.prev.prev.value = 1
assert dll.tail.prev.prev.prev = None
assert dll._size = 3

assert dll.pop() == 1
assert dll.head.prev = None
assert dll.head.value = 2
assert dll.head.next.value = 3
assert dll.head.next.next = None
assert dll.tail.next = None
assert dll.tail.value = 3
assert dll.tail.prev.value = 2
assert dll.tail.prev.prev = None
assert dll._size = 2

assert dll.pop() == 2
assert dll.head.prev = None
assert dll.head.value = 3
assert dll.head.next = None
assert dll.tail.next = None
assert dll.tail.value = 3
assert dll.tail.prev = None
assert dll._size = 1

assert dll.pop() == 3
assert dll.head = None
assert dll.tail = None
assert dll._size = 0
```

I know this seems like a lot, but you'll see why it's import as you try to get the tests to pass.

File: doubly_linked_list.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        
+        if node == self.tail:
+            self.tail = self.tail.prev
+            self._size -= 1
+            return
```
Comment (W-Ely) [inline doubly_linked_list.py]: > remove(val) will remove the first instance of val found in the list, starting from the head. ...

If the list looked like head[a]<-->[b]<-->[c]<-->[b] and .remove(b) was called, the list should look like a - c - b not a - b - c.  Please write a test that covers this case.

**Review by W-Ely — CHANGES_REQUESTED** [review]
This seems very close.  I think your tests are a bit insufficient to know it's doing the right thing.  A few points to look at.

Class Names are always UpperCamelCase.

In the resting state of the linked list (not mid operation) the following should be true.
- The tail's .next should always be `None`.
- The head's prev should always be `None`.

File: doubly_linked_list.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            if current.value == value:
+                return current
+            current = current.next
+        return None
+    
+    def __len__(self):
+        return self.size()
+
+    def size(self):
+        return self._size
```
Comment (W-Ely) [inline doubly_linked_list.py]: Remove this method, it's not in the instructions.

**Review by W-Ely — COMMENTED** [review]
Looks great, one small thing.

**Review by W-Ely — APPROVED** [review]

**Pull request merged**